File: testsuite/data/scores/sample_crowdsale/sample_crowdsale.py
from iconservice import *

TAG = 'SampleCrowdsale'


# An interface of token to give a reward to anyone who contributes
class TokenInterface(InterfaceScore):
    @interface
    def transfer(self, _to: Address, _value: int, _data: bytes = None):
        pass


class SampleCrowdsale(IconScoreBase):

    _ADDR_BENEFICIARY = 'addr_beneficiary'
    _ADDR_TOKEN_SCORE = 'addr_token_score'
    _FUNDING_GOAL = 'funding_goal'
    _AMOUNT_RAISED = 'amount_raised'
    _DEAD_LINE = 'dead_line'
    _PRICE = 'price'
    _BALANCES = 'balances'
    _JOINER_LIST = 'joiner_list'
    _FUNDING_GOAL_REACHED = 'funding_goal_reached'
    _CROWDSALE_CLOSED = 'crowdsale_closed'

    ONE_ICX = 10 ** 18

    @eventlog
    def CrowdsaleStarted(self, fundingGoal: int, deadline: int):
        pass

    @eventlog
    def CrowdsaleEnded(self):
        pass

    @eventlog(indexed=3)
    def FundTransfer(self, backer: Address, amount: int, is_contribution: bool):
        pass

    @eventlog(indexed=2)
    def GoalReached(self, recipient: Address, total_amount_raised: int):
        pass

    def __init__(self, db: IconScoreDatabase) -> None:
        super().__init__(db)

        self._addr_beneficiary = VarDB(self._ADDR_BENEFICIARY, db, value_type=Address)
        self._addr_token_score = VarDB(self._ADDR_TOKEN_SCORE, db, value_type=Address)
        self._funding_goal = VarDB(self._FUNDING_GOAL, db, value_type=int)
        self._amount_raised = VarDB(self._AMOUNT_RAISED, db, value_type=int)
        self._dead_line = VarDB(self._DEAD_LINE, db, value_type=int)
        self._price = VarDB(self._PRICE, db, value_type=int)
        self._balances = DictDB(self._BALANCES, db, value_type=int)
        self._joiner_list = ArrayDB(self._JOINER_LIST, db, value_type=Address)
        self._funding_goal_reached = VarDB(self._FUNDING_GOAL_REACHED, db, value_type=bool)
        self._crowdsale_closed = VarDB(self._CROWDSALE_CLOSED, db, value_type=bool)

    def on_install(self, _fundingGoalInIcx: int, _tokenScore: Address, _durationInBlocks: int) -> None:
        """
        Called when this SCORE first deployed.

        :param _fundingGoalInIcx: The funding goal of this crowdsale, in ICX
        :param _tokenScore: SCORE address of token that will be used for the rewards
        :param _durationInBlocks: the sale duration is given in number of blocks
        """
        super().on_install()

        Logger.debug(f'on_install: fundingGoalInIcx={_fundingGoalInIcx}', TAG)
        Logger.debug(f'on_install: tokenScore={_tokenScore}', TAG)
        Logger.debug(f'on_install: durationInBlocks={_durationInBlocks}', TAG)

        if _fundingGoalInIcx < 0:
            revert("Funding goal cannot be less than zero")

        if _durationInBlocks < 0:
            revert("Duration cannot be less than zero")

        # The exchange ratio to ICX is 1:1
        icx_cost_of_each_token = 1

        self._addr_beneficiary.set(self.msg.sender)
        self._addr_token_score.set(_tokenScore)
        self._funding_goal.set(_fundingGoalInIcx * self.ONE_ICX)
        self._dead_line.set(self.block.height + _durationInBlocks)
        price = int(icx_cost_of_each_token)
        self._price.set(price)

        self._funding_goal_reached.set(False)
        self._crowdsale_closed.set(True)  # Crowdsale closed by default

    def on_update(self) -> None:
        super().on_update()

    @external
    def tokenFallback(self, _from: Address, _value: int, _data: bytes):
        """
        Implements `tokenFallback` in order for the SCORE
        to receive initial tokens to reward to the contributors
        """

        # Checks if the caller is a Token SCORE address that this SCORE is interested in.
        if self.msg.sender != self._addr_token_score.get():
            revert("Unknown token address")

        # Depositing tokens can only be done by owner
        if _from != self.owner:
            revert("Invalid sender")

        if _value < 0:
            revert("Depositing value cannot be less than zero")

        # start Crowdsale hereafter
        self._crowdsale_closed.set(False)
        Logger.debug(f'tokenFallback: token supply = "{_value}"', TAG)
        self.CrowdsaleStarted(self._funding_goal.get(), self._dead_line.get())

    @payable
    def fallback(self):
        """
        Called when anyone sends funds to the SCORE.
        This SCORE regards it as a contribution.
        """
        if self._crowdsale_closed.get():
            revert('Crowdsale is closed.')

        # Accepts the contribution
        amount = self.msg.value
        self._balances[self.msg.sender] = self._balances[self.msg.sender] + amount
        self._amount_raised.set(self._amount_raised.get() + amount)
        value = int(amount / self._price.get())
        data = b'called from Crowdsale'

        # Gives tokens to the contributor as a reward
        token_score = self.create_interface_score(self._addr_token_score.get(), TokenInterface)
        token_score.transfer(self.msg.sender, value, data)

        if self.msg.sender not in self._joiner_list:
            self._joiner_list.put(self.msg.sender)

        self.FundTransfer(self.msg.sender, amount, True)
        Logger.debug(f'FundTransfer({self.msg.sender}, {amount}, True)', TAG)

    @external(readonly=True)
    def totalJoinerCount(self) -> int:
        """
        Returns the number of contributors.

        :return: the number of contributors
        """
        return len(self._joiner_list)

    def _after_dead_line(self) -> bool:
        # Checks if it has been reached to the deadline block
        Logger.debug(f'after_dead_line: block.height = {self.block.height}', TAG)
        Logger.debug(f'after_dead_line: dead_line()  = {self._dead_line.get()}', TAG)
        return self.block.height >= self._dead_line.get()

    @external
    def checkGoalReached(self):
        """
        Checks if the goal has been reached and ends the campaign.
        """
        if self._after_dead_line():
            if not self._crowdsale_closed.get():
                self._crowdsale_closed.set(True)
                self.CrowdsaleEnded()

            if self._amount_raised.get() >= self._funding_goal.get():
                self._funding_goal_reached.set(True)
                self.GoalReached(self._addr_beneficiary.get(), self._amount_raised.get())
                Logger.debug(f'Goal reached!', TAG)

    @external
    def safeWithdrawal(self):
        """
        Withdraws the funds.

        If the funding goal has been reached, sends the entire amount to the beneficiary.
        If the goal was not reached, each contributor can withdraw the amount they contributed.
        """
        if self._after_dead_line():
            # each contributor can withdraw the amount they contributed if the goal was not reached
            if not self._funding_goal_reached.get():
                amount = self._balances[self.msg.sender]
                self._balances[self.msg.sender] = 0
                if amount > 0:
                    if self.icx.send(self.msg.sender, amount):
                        self.FundTransfer(self.msg.sender, amount, False)
                        Logger.debug(f'FundTransfer({self.msg.sender}, {amount}, False)', TAG)
                    else:
                        self._balances[self.msg.sender] = amount

            # The sales target has been met. Owner can withdraw the contribution.
            if self._funding_goal_reached.get() and self._addr_beneficiary.get() == self.msg.sender:
                if self.icx.send(self._addr_beneficiary.get(), self._amount_raised.get()):
                    self.FundTransfer(self._addr_beneficiary.get(), self._amount_raised.get(), False)
                    Logger.debug(f'FundTransfer({self._addr_beneficiary.get()},'
                                 f'{self._amount_raised.get()}, False)', TAG)
                    # reset amount_raised
                    self._amount_raised.set(0)
                else:
                    # if the transfer to beneficiary fails, unlock contributors balance
                    Logger.debug(f'Failed to send to beneficiary!', TAG)
                    self._funding_goal_reached.set(False)

File: javaee/samples/crowdsale/src/example/SampleCrowdsale.java
/*
 * Copyright 2019 ICON Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package example;

import score.Address;
import score.Context;
import score.DictDB;
import score.VarDB;
import score.annotation.EventLog;
import score.annotation.External;
import score.annotation.Payable;

import java.math.BigInteger;

public class SampleCrowdsale
{
    private static final BigInteger ONE_ICX = new BigInteger("1000000000000000000");
    private final Address beneficiary;
    private final Address tokenScore;
    private final BigInteger fundingGoal;
    private final long deadline;
    private boolean fundingGoalReached;
    private boolean crowdsaleClosed;
    private final DictDB<Address, BigInteger> balances;
    private final VarDB<BigInteger> amountRaised;

    public SampleCrowdsale(BigInteger _fundingGoalInIcx, Address _tokenScore, BigInteger _durationInBlocks) {
        // some basic requirements
        Context.require(_fundingGoalInIcx.compareTo(BigInteger.ZERO) >= 0);
        Context.require(_durationInBlocks.compareTo(BigInteger.ZERO) >= 0);

        this.beneficiary = Context.getCaller();
        this.fundingGoal = ONE_ICX.multiply(_fundingGoalInIcx);
        this.tokenScore = _tokenScore;
        this.deadline = Context.getBlockHeight() + _durationInBlocks.longValue();

        this.fundingGoalReached = false;
        this.crowdsaleClosed = true; // Crowdsale closed by default

        this.balances = Context.newDictDB("balances", BigInteger.class);
        this.amountRaised = Context.newVarDB("amountRaised", BigInteger.class);
    }

    /*
     * Receives initial tokens to reward to the contributors.
     */
    @External
    public void tokenFallback(Address _from, BigInteger _value, byte[] _data) {
        // check if the caller is a token SCORE address that this SCORE is interested in
        Context.require(Context.getCaller().equals(this.tokenScore));

        // depositing tokens can only be done by owner
        Context.require(Context.getOwner().equals(_from));

        // value should be greater than zero
        Context.require(_value.compareTo(BigInteger.ZERO) >= 0);

        // start Crowdsale hereafter
        Context.require(this.crowdsaleClosed);
        this.crowdsaleClosed = false;
        // emit eventlog
        CrowdsaleStarted(this.fundingGoal, this.deadline);
    }

    /*
     * Called when anyone sends funds to the SCORE and that funds would be regarded as a contribution.
     */
    @Payable
    public void fallback() {
        // check if the crowdsale is closed
        Context.require(!this.crowdsaleClosed);

        Address _from = Context.getCaller();
        BigInteger _value = Context.getValue();
        Context.require(_value.compareTo(BigInteger.ZERO) > 0);

        // accept the contribution
        BigInteger fromBalance = safeGetBalance(_from);
        this.balances.set(_from, fromBalance.add(_value));

        // increase the total amount of funding
        BigInteger amountRaised = safeGetAmountRaised();
        this.amountRaised.set(amountRaised.add(_value));

        // give tokens to the contributor as a reward
        byte[] _data = "called from Crowdsale".getBytes();
        Context.call(this.tokenScore, "transfer", _from, _value, _data);
        // emit eventlog
        FundTransfer(_from, _value, true);
    }

    /*
     * Checks if the goal has been reached and ends the campaign.
     */
    @External
    public void checkGoalReached() {
        if (afterDeadline()) {
            if (!this.crowdsaleClosed) {
                this.crowdsaleClosed = true;
                // emit eventlog
                CrowdsaleEnded();
            }
            BigInteger amountRaised = safeGetAmountRaised();
            if (amountRaised.compareTo(this.fundingGoal) >= 0) {
                this.fundingGoalReached = true;
                // emit eventlog
                GoalReached(this.beneficiary, amountRaised);
            }
        }
    }

    /*
     * Withdraws the funds safely.
     *
     *  - If the funding goal has been reached, sends the entire amount to the beneficiary.
     *  - If the goal was not reached, each contributor can withdraw the amount they contributed.
     */
    @External
    public void safeWithdrawal() {
        if (afterDeadline()) {
            Address _from = Context.getCaller();

            // each contributor can withdraw the amount they contributed if the goal was not reached
            if (!this.fundingGoalReached) {
                BigInteger amount = safeGetBalance(_from);
                if (amount.compareTo(BigInteger.ZERO) > 0) {
                    // set their balance to ZERO first before transferring the amount to prevent reentrancy attack
                    this.balances.set(_from, BigInteger.ZERO);
                    // transfer the icx back to them
                    Context.transfer(_from, amount);
                    // emit eventlog
                    FundTransfer(_from, amount, false);
                }
            }

            // owner can withdraw the contribution since the sales target has been met.
            if (this.fundingGoalReached && this.beneficiary.equals(_from)) {
                BigInteger amountRaised = safeGetAmountRaised();
                if (amountRaised.compareTo(BigInteger.ZERO) > 0) {
                    // transfer the funds to beneficiary
                    Context.transfer(this.beneficiary, amountRaised);
                    // emit eventlog
                    FundTransfer(this.beneficiary, amountRaised, false);
                    // reset amountRaised
                    this.amountRaised.set(BigInteger.ZERO);
                }
            }
        }
    }

    private BigInteger safeGetBalance(Address owner) {
        return this.balances.getOrDefault(owner, BigInteger.ZERO);
    }

    private BigInteger safeGetAmountRaised() {
        return this.amountRaised.getOrDefault(BigInteger.ZERO);
    }

    private boolean afterDeadline() {
        // checks if it has been reached to the deadline block
        return Context.getBlockHeight() >= this.deadline;
    }

    @EventLog
    private void CrowdsaleStarted(BigInteger fundingGoal, long deadline) {}

    @EventLog
    private void CrowdsaleEnded() {}

    @EventLog(indexed=3)
    private void FundTransfer(Address backer, BigInteger amount, boolean isContribution) {}

    @EventLog(indexed=2)
    private void GoalReached(Address recipient, BigInteger totalAmountRaised) {}
}
